### PR TITLE
kotlinx.serialization to mojang serialization bridge

### PR DESCRIPTION
**Pull request by Cypher121**
Adds a new feature allowing the following:

* Conversion of KSerializers (generated by Serializable annotations) to Codecs
* Usage of pre-existing codecs for generating new ones
* Extensive format configuration on a per-type or per-codec basis.

User and implementer documentation can be found within the package folder in serialization.md

Some things that may warrant more work:
* Documentation, either overview in .md or KDoc and comments on individual issues
* Making tests run in an actual test engine. Requires resolving loom dependency issues (probably #45).
* Decoder/Encoder state management. May be a sane way around ElementOptions hell.
* More tests. Even more tests. MORE. Any ideas welcome, especially those with test code in them.

### CODE REVIEW

**Review by Oliver-makes-code — APPROVED** [review]
LGTM

**Review by PotatoPresident — APPROVED** [review]
This is awesome! Definitely going to migrate my codecs to this

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/serialization/internal/SerializerCodec.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import com.mojang.serialization.Codec
+import com.mojang.serialization.DataResult
+import com.mojang.serialization.DynamicOps
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.modules.SerializersModule
+import org.quiltmc.qkl.wrapper.minecraft.serialization.internal.decoder.DynamicDecoder
+import org.quiltmc.qkl.wrapper.minecraft.serialization.internal.encoder.DynamicEncoder
+import org.quiltmc.qkl.wrapper.minecraft.serialization.options.CodecOptions
+
+internal class SerializerCodec<A>(
```
Comment (SilverAndro) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/serialization/internal/SerializerCodec.kt:29]: This class should have toString implemented for debugging purposes, as is it just returns the jvm representation

image

**Review by SilverAndro — CHANGES_REQUESTED** [review]

**Comment by SilverAndro** [comment]
Seems to work in my mod (REA usage specifically tested) so other than the minor thing about toString im happy to approve, absolutely incredible work!

**Review by SilverAndro — APPROVED** [review]

**Review by PotatoPresident — APPROVED** [review]

**Review by NoComment1105 — APPROVED** [review]

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -9,6 +9,7 @@ qsl = "3.0.0-beta.16+1.19.2"
 quilt_loader = "0.17.4"
 quilt_mappings = "1.19.2+build.1"
 serialization = "1.4.0"
+serialization_plugin = "1.7.10" # usually same as kotlin but can lag behind
```
Comment (Cypher121) [inline gradle/libs.versions.toml:12]:
```suggestion
serialization_plugin = "1.7.20" # usually same as kotlin but can lag behind
```

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -9,6 +9,7 @@ qsl = "3.0.0-beta.16+1.19.2"
 quilt_loader = "0.17.4"
 quilt_mappings = "1.19.2+build.1"
 serialization = "1.4.0"
```
Comment (Cypher121) [inline gradle/libs.versions.toml:11]:
```suggestion
serialization = "1.4.1"
```

**Pull request merged**